File: PythonAlgorithmInterView/문자열조작/그룹애너그램.py
import collections

# rock solution
def groupAnagrams(self, strs: list[str]) -> list[list[str]]:

    dict = collections.defaultdict(list)

    for str in strs :
        str_list = list(str)
        str_list.sort()
        dict[''.join(str_list)].append(str)
        dict[''.join(str_list)].sort()

    return list(dict.values())

# sang solution
def groupAnagrams1(self, strs: list[str]) -> list[list[str]]:

    dict = collections.defaultdict(list)

    for str in strs :
        dict[''.join(sorted(str))].append(str)

    return list(dict.values())

if __name__ == '__main__':
    strs = ["eat", "tea", "tan", "ate", "nat", "bat"]

    print(groupAnagrams1(None, strs))

    max()

File: PythonAlgorithmInterView/문자열조작/그룹애너그램.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 * 45번 문제.
 */
 var groupAnagrams = function(strs) {
    let argument = {}

    strs.forEach( str => {
        argument[str.split('').sort().toString()] = []
    })

    strs.forEach( str => {
        argument[str.split('').sort().toString()] = argument[str.split('').sort().toString()].concat(str).sort();
    })

    return Object.values(argument)
};

strs = ["eat","tea","tan","ate","nat","bat"]


console.log(groupAnagrams(strs))